### PR TITLE
Trovo OAuth provider

### LINKED ISSUE #642: Trovo Support
**Is your feature request related to a problem? Please describe.**
Trovo is a Stream platform that supports [OAuth](https://developer.trovo.live/docs/APIs.html#_3-authentication) and it would be great to support in ASP.NET.

**Describe the solution you'd like**
Added support to Trovo Authentication

**Describe alternatives you've considered**
I will try to implement my own solution for this integration.

### PR DESCRIPTION
**Pull request by serber**
Created Trovo Support for issue #642

### CODE REVIEW

File: docs/trovo.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Integrating the Trovo Provider
```
Comment (martincostello) [inline docs/trovo.md:1]: We can keep this anyway, but we don't generally bother adding documentation for providers that don't have any custom settings.

File: src/AspNet.Security.OAuth.Trovo/AspNet.Security.OAuth.Trovo.csproj [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
+  </PropertyGroup>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Trovo/AspNet.Security.OAuth.Trovo.csproj]: Could you also add this please?

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/f51b62b4b63406ef04198bb780084c2fedabc6d7/src/AspNet.Security.OAuth.Twitter/AspNet.Security.OAuth.Twitter.csproj#L4

File: src/AspNet.Security.OAuth.Trovo/TrovoAuthenticationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using AspNet.Security.OAuth.Trovo;
+
+namespace Microsoft.Extensions.DependencyInjection;
+
+/// <summary>
+/// Extension methods to add  authentication capabilities to an HTTP application pipeline.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Trovo/TrovoAuthenticationExtensions.cs]:
```suggestion
/// Extension methods to add Trovo authentication capabilities to an HTTP application pipeline.
```

File: src/AspNet.Security.OAuth.Trovo/TrovoAuthenticationConstants.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        public const string ChannelId = "urn:trovo:channelid";
+        public const string ProfilePic = "urn:trovo:profilepic";
+    }
+
+    public static class Headers
+    {
+        public const string ClientId = "client-id";
+        public const string Authorization = "OAuth";
+    }
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Trovo/TrovoAuthenticationConstants.cs]: I don't think this needs to be public. `Authorization` is only used in one place, `ClientId` could just be a private constant in the handler.

File: src/AspNet.Security.OAuth.Trovo/TrovoAuthenticationConstants.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+namespace AspNet.Security.OAuth.Trovo;
+
+/// <summary>
+/// Contains constants specific to the <see cref="TrovoAuthenticationHandler"/>.
+/// </summary>
+public class TrovoAuthenticationConstants
+{
+    public const string GrantType = "authorization_code";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Trovo/TrovoAuthenticationConstants.cs]: Similarly here.

**Review by martincostello — COMMENTED** [review]
Just a few minor comments.

**Comment by martincostello** [comment]
Thanks for your contribution @serber. A standard question for a new provider, can you confirm you've also validated it works as expected against the real Trovo service?

**Comment by serber** [comment]
@martincostello, yes i test this code in real app.
Here fiddler logs

**Getting a token**
```
POST https://open-api.trovo.live/openplatform/exchangetoken HTTP/1.1
Host: open-api.trovo.live
client-id: 9d7d9e8b7673faa226d135cdfbce50cf
Accept: application/json
User-Agent: Microsoft ASP.NET Core OAuth handler
traceparent: 00-6051613d19c65a66779bedbba7e55dc0-71a931362b15cab4-00
Content-Type: application/json; charset=utf-8
Content-Length: 181

{"redirect_uri":"https://***/signin-trovo","code":"***","client_secret":"***","grant_type":"authorization_code"}


HTTP/1.1 200 OK
Date: Fri, 21 Jan 2022 19:32:51 GMT
Content-Type: application/json
Content-Length: 143
Connection: keep-alive
grpc-metadata-content-type: application/grpc
grpc-metadata-date: Fri, 21 Jan 2022 19:32:51 GMT
grpc-metadata-server: envoy
grpc-metadata-server-ip: 172.19.9.206
grpc-metadata-x-envoy-upstream-service-time: 27
x-ratelimit-limit: 1200
x-ratelimit-remaining: 1195
x-ratelimit-reset: 1642793580
x-envoy-upstream-service-time: 45
server: istio-envoy

{"access_token":"***","token_type":"bearer","expires_in":14400,"refresh_token":"***"}
```

**Getting user information**
```
GET https://open-api.trovo.live/openplatform/getuserinfo HTTP/1.1
Host: open-api.trovo.live
client-id: 9d7d9e8b7673faa226d135cdfbce50cf
Accept: application/json
Authorization: OAuth 5a792987dd0b9d928a55ac57abc89b1d
User-Agent: Microsoft ASP.NET Core OAuth handler
traceparent: 00-6051613d19c65a66779bedbba7e55dc0-020906356f588cc6-00


HTTP/1.1 200 OK
Date: Fri, 21 Jan 2022 19:32:52 GMT
Content-Type: application/json
Content-Length: 190
Connection: keep-alive
grpc-metadata-content-type: application/grpc
grpc-metadata-date: Fri, 21 Jan 2022 19:32:52 GMT
grpc-metadata-server: envoy
grpc-metadata-server-ip: 172.19.5.151
grpc-metadata-x-envoy-upstream-service-time: 11
x-ratelimit-limit: 1200
x-ratelimit-remaining: 1194
x-ratelimit-reset: 1642793580
x-envoy-upstream-service-time: 58
server: istio-envoy

{"userId":"***","userName":"***","nickName":"***","email":"***","profilePic":"https://headicon.trovo.live/default/6.png","info":"","channelId":"***"}
```

**Review by martincostello — APPROVED** [review]

**Pull request merged**